proposals/NNNN-isolated-synchronous-deinit.md

```diff
@@ -0,0 +1,386 @@
+
+In cases when `deinit` belongs to a subclass of `UIView` or `UIViewController` which are known to call deinitializer on the main thread, developers may be tempted to silence the diagnostic by adopting `@unchecked Sendable` in types that are not actually  sendable. This undermines concurrency checking by the compiler, and may lead to data races when using incorrectly marked types in other places.
+
+## Proposed solution
+
+... is to allow execution of `deinit` and object deallocation to be the scheduled on the executor, if needed.
+
+Let's consider [examples from SE-0327](https://github.com/apple/swift-evolution/blob/main/proposals/0327-actor-initializers.md#data-races-in-deinitializers):
+
+In case of several instances with shared data isolated on common actor problem is completely eliminated:
+
+```swift
+class NonSendableAhmed { 
+  var state: Int = 0
+}
+
+@MainActor
+class Maria {
+  let friend: NonSendableAhmed
+
+  init() {
+    self.friend = NonSendableAhmed()
+  }
+
+  init(sharingFriendOf otherMaria: Maria) {
+    // While the friend is non-Sendable, this initializer and
+    // and the otherMaria are isolated to the MainActor. That is,
+    // they share the same executor. So, it's OK for the non-Sendable value
+    // to cross between otherMaria and self.
+    self.friend = otherMaria.friend
+  }
+
+  deinit {
+    // Used to be a potential data race.
+    // Now deinit is also isolated on the MainActor.
+    // So this code is perfectly correct.
+    friend.state += 1
+  }
+}
+
+func example() async {
+  let m1 = await Maria()
+  let m2 = await Maria(sharingFriendOf: m1)
+  doSomething(m1, m2)
+} 
+```
+
+In case of escaping self, race condition is eliminated but problem of escaping `self` remains. This problem exists for synchronous code as well and is orthogonal to the concurrency features.
+
+```swift
+actor Clicker {
+  var count: Int = 0
+
+  func click(_ times: Int) {
+    for _ in 0..<times {
+      self.count += 1 
+    }
+  }
+
+  deinit {
+    let old = count
+    let moreClicks = 10000
+    
+    Task { await self.click(moreClicks) } // ❌ This WILL keep `self` alive after the `deinit`!
+
+    for _ in 0..<moreClicks {
+        // No data race.
+        // Actor job created by the task is either
+        // not created yet or is waiting in the queue.
+      self.count += 1 
+    }
+
+    assert(count == old + moreClicks) // Always works
+    assert(count == old + 2 * moreClicks) // Always fails
+  }
+}
+```
+
+## Detailed design
+
+### Runtime
+
+Proposal introduces new runtime function that is used to schedule task-less block of synchronous code on the executor by wrapping it into an ad hoc task. It does not do any reference counting and can be safely used even with references that were released for the last time but not deallocated yet.
```

Comment:
I was looking for a word that would mean "conceptually a closure, but technically not a Swift closure". I didn't want to use term `closure` here, because runtime function does not use closure types to avoid reference counting.

Reply:
Right, I see the difficulty and agree that 'closure' is wrong here too. Would "task-less _job_" appropriately capture this? IIRC that's the terminology we use for each unit of a task that runs synchronously until we hit a suspension point, but I don't know if it would be appropriate to call this a job as well.